packages/csv-parser-state/src/main.js

```diff
@@ -0,0 +1,190 @@
+/* @flow */
+import csv from 'csv-parser'
+import highland from 'highland'
+import JSONStream from 'JSONStream'
+import memoize from 'lodash.memoize'
+import { unflatten } from 'flat'
+import { createClient } from '@commercetools/sdk-client'
+import { createRequestBuilder } from '@commercetools/api-request-builder'
+import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
+import {
+  createAuthMiddlewareForClientCredentialsFlow,
+  createAuthMiddlewareWithExistingToken,
+} from '@commercetools/sdk-middleware-auth'
+import { createUserAgentMiddleware } from '@commercetools/sdk-middleware-user-agent'
+import type {
+  ApiConfigOptions,
+  ConstructorOptions,
+  CsvOptions,
+  LoggerOptions,
+  StateWithStringTransitions,
+  StateWithUnresolvedTransitions,
+} from 'types/state'
+import type { Client, SuccessResult } from 'types/sdk'
+import pkg from '../package.json'
+
+export default class CsvParserState {
+  // Set flowtype annotations
+  accessToken: string
+  apiConfig: ApiConfigOptions
+  client: Client
+  continueOnProblems: boolean
+  csvConfig: CsvOptions
+  logger: LoggerOptions
+  _rowIndex: number
+  _fetchStates: Function
+
+  constructor(
+    {
+      apiConfig,
+      csvConfig,
+      continueOnProblems = false,
+      accessToken,
+    }: ConstructorOptions,
+    logger: LoggerOptions
+  ) {
+    this.apiConfig = apiConfig
+    this.accessToken = accessToken
+
+    const defaultOptions = {
+      delimiter: ',',
+      multiValueDelimiter: ';',
+    }
+    this.csvConfig = { ...defaultOptions, ...csvConfig }
+    this.continueOnProblems = continueOnProblems
+    this.logger = logger || {
+      error: () => {},
+      warn: () => {},
+      info: () => {},
+      debug: () => {},
+    }
+
+    this._rowIndex = 0
+  }
+
+  parse(input: stream$Readable, output: stream$Writable) {
+    this.logger.info('Starting conversion')
+    highland(input)
+      .through(csv({ separator: this.csvConfig.delimiter, strict: true }))
+      .map(CsvParserState._removeEmptyFields)
+      .map(CsvParserState._parseInitialToBoolean)
+      .map(unflatten)
+      .map(state => this._mapMultiValueFieldsToArray(state))
+      .flatMap(state => highland(this._transformTransitions(state)))
```

Comment:
Nitpick: we dont need the wrapping fn.

Reply:
We do. The `.flatmap()` expects a stream to be passed to the callback and because `. _transformTransitions()` returns a promise, the `highland()` wrapper transforms the contents of the resolved (or rejected promise to stream chunks